addons/account/models/account_move.py

```diff
@@ -1,349 +1,1874 @@
+
+            # Copy currency.
+            if self.currency_id != self.invoice_vendor_bill_id.currency_id:
+                self.currency_id = self.invoice_vendor_bill_id.currency_id
+
+            # Reset
+            self.invoice_vendor_bill_id = False
+            self._recompute_dynamic_lines()
+
+    @api.onchange('type')
+    def _onchange_type(self):
+        ''' Onchange made to filter the partners depending of the type. '''
+        if self.type in ('out_invoice', 'out_refund', 'out_receipt'):
+            return {'domain': {'partner_id': [('customer', '=', True)]}}
+        elif self.type in ('in_invoice', 'in_refund', 'in_receipt'):
+            return {'domain': {'partner_id': [('supplier', '=', True)]}}
+
+    @api.onchange('invoice_line_ids')
+    def _onchange_invoice_line_ids(self):
+        current_invoice_lines = self.line_ids.filtered(lambda line: not line.exclude_from_invoice_tab)
+        others_lines = self.line_ids - current_invoice_lines
+        if others_lines and current_invoice_lines - self.invoice_line_ids:
+            others_lines[0].recompute_tax_line = True
+        self.line_ids = others_lines + self.invoice_line_ids
+        self._onchange_recompute_dynamic_lines()
+
+    @api.onchange('line_ids', 'invoice_payment_term_id', 'invoice_date_due', 'invoice_cash_rounding_id', 'invoice_vendor_bill_id')
+    def _onchange_recompute_dynamic_lines(self):
+        self._recompute_dynamic_lines()
+
+    @api.model
+    def _build_tax_line_groupby_key(self, line, tax_repartition_line, tax_ids, tag_ids):
+        ''' Create a key used to build a tax mapping between the base lines and the tax lines.
+        This is necessary to aggregate the tax amounts depending of some fields.
+
+        /!\ As line could be a base line or a tax line, only common field could be used to build the key.
+        For example, the tax_ids field could not be used directly from the line as this field has a different
+        purpose on base / tax lines. This is the reason why some field are passed as independent parameters.
+
+        :param line:                    A base / tax account.move.line.
+        :param tax_repartition_line:    An account.tax.repartition.line.
+        :param taxes:                   An account.tax recordset.
+        :param tags:                    An account.tag recordset.
+        :return:                        A list to be hashed representing a key in the taxes_map.
+        '''
+        tax = tax_repartition_line.invoice_tax_id or tax_repartition_line.refund_tax_id
+        return [
+            tax_repartition_line.id,
+            tax_repartition_line.account_id.id or line.account_id.id,
+            line.currency_id.id,
+            tax.analytic and tuple(line.analytic_tag_ids.ids),
+            tax.analytic and line.analytic_account_id.id,
+            tuple(tax_ids),
+            tuple(tag_ids),
+        ]
+
+    @api.model
+    def _get_tax_line_values_from_base_lines(self, base_lines):
```

Comment:
based on the docstring, I would have expected a tax as parameter, or this method defined on a tax recordset...

what if the tax hasn't its `analytic` field set to True?

Reply:
:boom: